# Deep cfNA #


**Deep_cfNA** is a classifier for recognizing RNA and DNA fragments in TGIRT-seq plasma nucleotide sequencing libraries. TGIRTs (Thermostable group II intron reverse transcriptase) are both RNA-dependent and DNA-dependent DNA polymerases, we have demonstrated using TGIRTs for sequencing [cell-free DNA](https://www.nature.com/articles/s41598-017-09064-w) or [cell-free RNA](https://www.ncbi.nlm.nih.gov/pubmed/26554030) from human blood plasma. However, it has been a debate whether some of the fragments in TIGRT-seq plasma RNA-seq libraries were DNA contaminates or not. As it has been [proven](https://www.ncbi.nlm.nih.gov/pubmed/26771485) plasma cell-free DNA are nucleosome-bound DNA or transcription factor-bound and these DNA fragments should have some [nucleotide patterns](https://academic.oup.com/bioinformatics/article/34/10/1705/4796955), we hypothesize a convolutional neural network (CNN) can be used to pick up nucleotide patterns/signals/motifs for identifying DNA contaminants in TGIRT-seq of plasma nucleotides. 

There are several models out there using convolutional neural network (CNN) for identifying transcription factor binding sites or nucelosome binding sites across the genoms, such as [DeepBind](https://www.nature.com/articles/nbt.3300), [LeNup](https://academic.oup.com/bioinformatics/article/34/10/1705/479695) and [DanQ](https://www.ncbi.nlm.nih.gov/pubmed/27084946). Among these models, [DanQ](https://www.ncbi.nlm.nih.gov/pubmed/27084946) has presented an interesting novel hybrdi convolutional and recurrent neural network (RNN) model. This model utilized a convolutional layer for picking up motifs, and a bidirectional long short term memory (LSTM) layer to identify important spatial arrangement of these motifs. As these models have proven deep learning using DNA/RNA fragments as inputs can accurately predict DNA-binding protein preferred binding sites, we are interesting at using similar technique to pick up DNA contamination fragments from plasma RNA libraries generated by TGIRT-seq.

To test whether deep neural network can be used in identifying DNA contaminants in TGIRT-seq plasma RNA-seq librarys, we built a model with similar architecture as [DanQ](https://github.com/uci-cbcl/DanQ/blob/master/DanQ_train.py) and tested on TGIRT-seq data from (1) alkaline-hydrolysis plasma nucleotides, and (2) DNase treated plasma nucleotides.




