# Deep cfNA #

**Deep_cfNA** is a classifier for recognizing RNA and DNA fragments in TGIRT-seq plasma nucleotide sequencing libraries. TGIRTs (Thermostable group II intron reverse transcriptase) are both RNA-dependent and DNA-dependent DNA polymerases, we have demonstrated using TGIRTs for sequencing [cell-free DNA](https://www.nature.com/articles/s41598-017-09064-w) or [cell-free RNA](https://www.ncbi.nlm.nih.gov/pubmed/26554030) from human blood plasma. However, it has been a debate whether some of the fragments in TIGRT-seq plasma RNA-seq libraries were DNA contaminates or not. As it has been [proven](https://www.ncbi.nlm.nih.gov/pubmed/26771485) plasma cell-free DNA are nucleosome-bound or transcription factor-bound, thus they should have some [motif patterns](https://academic.oup.com/bioinformatics/article/34/10/1705/4796955), we hypothesize a convolutional neural network (CNN) can be used to pick up subtle nucleotide patterns/signals/motifs for identifying DNA contaminants in TGIRT-seq of plasma nucleotides. 

There are several deep learning models out there using convolutional neural network (CNN) for identifying transcription factor binding sites or nucelosome binding sites across the genoms, such as [DeepBind](https://www.nature.com/articles/nbt.3300), [LeNup](https://academic.oup.com/bioinformatics/article/34/10/1705/479695) and [DanQ](https://www.ncbi.nlm.nih.gov/pubmed/27084946). Among these models, [DanQ](https://www.ncbi.nlm.nih.gov/pubmed/27084946) has presented an interesting novel hybrid convolutional and recurrent neural network (RNN) model. This model utilized a convolutional layer for picking up motifs, and a bidirectional long short term memory (LSTM) layer to identify important spatial arrangement of these motifs on both Watson and Crick strands. As these models have proven deep learning using one-hot encoded DNA/RNA fragments alone as inputs can accurately predict DNA-binding protein preferred binding sites, we are interesting at using similar technique to pick up DNA contamination fragments from plasma RNA libraries generated by TGIRT-seq.

To test whether deep neural network can be used in identifying DNA contaminants in TGIRT-seq plasma RNA-seq librarys, we built a model with similar architecture as [DanQ](https://github.com/uci-cbcl/DanQ/blob/master/DanQ_train.py) and tested on TGIRT-seq data from (1) alkaline-hydrolysis plasma nucleotides, and (2) DNase treated plasma nucleotides.


# Usage #

Deep_cfNA has two sub-commands: ```train``` and ```predict```. 

### train  ###

* ```train```: for training a new model, required: 

    * a genome fasta file and, 
    * a fragment bed file, the last columns must be ```RNA``` or ```DNA```, an example:
```
chr1    10027   10216   frag_1    0       +       DNA
chr1    10045   10105   frag_2    0       +       RNA
```

* A validation bed file can also be supplied to give validation statistics: 
    * AUROC: area under receiver opereator curve  
    * Precision
    * Recall: false positive rate
    * F1 score: harmonic mean of Precision and recall

* if ```--model_prefix``` is supplied, the neural network (implemented by [Keras](https://keras.io/)) can be saved


```
usage: deep_cfNA train [-h] --train_bed TRAIN_BED --genome GENOME
                             [--validation_bed VALIDATION_BED]
                             [--model_prefix MODEL_PREFIX]

optional arguments:
  -h, --help            show this help message and exit
  --train_bed TRAIN_BED
                        bed file for fragments: For each record in bed file,
                        extract the sequence, and center it fill up both sides
                        to length of (seq_length) with Ns. Bed files need
                        these columns: 1. chrom 2. start 3. end 4. 5. 6.
                        strand 7. label: (DNA or RNA)
  --genome GENOME       genome fasta file (must have faidx)
  --validation_bed VALIDATION_BED
                        bed file for validation
  --model_prefix MODEL_PREFIX
                        where to save the model
```


### predict ###

* ```predict```: for predicting labels, required:
    * a genome fasta file
    * a fragment bed file

* if ```--model_prefix``` can be used to supply a pre-trained model, otherwise, it will use the pre-trained model packaged within **Deep_cfNA**

* Output is send to stdout, where the input bed file is copied with predicted labels as an extra columns at the end of each line


```
usage: deep_cfNA predict [-h] --inbed INBED --genome GENOME
                               [--model_prefix MODEL_PREFIX]

optional arguments:
  -h, --help            show this help message and exit
  --inbed INBED         bed file for fragments: For each record in bed file,
                        extract the sequence, and center it fill up both sides
                        to length of (seq_length) with Ns. Bed files need
                        these columns: 1. chrom 2. start 3. end 4. 5. 6.
                        strand 7. label: (DNA or RNA)
  --genome GENOME       genome fasta file (must have faidx)
  --model_prefix MODEL_PREFIX
```


## Requirements ##
* [Keras](https://keras.io/) as deep learning framework
* [TensorFlow](https://www.tensorflow.org/) as backend
* [sequencing_tools](https://github.com/wckdouglas/sequencing_tools) for DNA sequence manipulations

## Validation ##
[Validation] Precision: 0.851
[Validation] Recall: 0.926
[Validation] F1: 0.887
[Validation] AUROC: 0.944
